File: installers/macos/menubar/Sources/LoombreIPCKit/Status.swift
// SPDX-License-Identifier: AGPL-3.0-only
// Loombre :: LoombreIPCKit/Status.swift
//
// Mirrors packages/controller-ipc/src/status.ts — GET /ipc/v1/status.

import Foundation

public struct IPCStatusResponse: Codable, Equatable {
    public let ipcContractVersion: Int
    public let server: IPCProcessInfo
    public let worker: IPCProcessInfo
    /// nil while the server is not in a state that serves the web client.
    public let webUrl: String?
    public let provisioning: ProvisioningStatus

    public init(
        ipcContractVersion: Int,
        server: IPCProcessInfo,
        worker: IPCProcessInfo,
        webUrl: String?,
        provisioning: ProvisioningStatus
    ) {
        self.ipcContractVersion = ipcContractVersion
        self.server = server
        self.worker = worker
        self.webUrl = webUrl
        self.provisioning = provisioning
    }

    /// True when this response's contract version is one this client
    /// build understands. A newer server (higher ipcContractVersion) is
    /// NOT necessarily a hard error — additive changes within v1 are
    /// expected (contract-version.ts's own comment) — but the menubar
    /// surfaces a notice either way per the mission brief ("version +
    /// contract-version mismatch notice").
    public var matchesClientContractVersion: Bool {
        ipcContractVersion == controllerIpcContractVersion
    }
}

File: packages/controller-ipc/src/status.ts
// SPDX-License-Identifier: AGPL-3.0-only
// Loombre :: packages/controller-ipc/src/status.ts
//
// GET /ipc/v1/status — server+worker state, versions, web URL, and the
// embedded-PG ProvisioningStatus passthrough. This is the ONLY
// cross-package import this whole package makes (the brief this package
// was built against is explicit that it must be the only one): reusing
// @loombre/provisioning's ProvisioningStatus type + schema verbatim rather
// than re-declaring a parallel copy that could drift.
//
// Also carries VersionInfo (ipcContractVersion) — the version-negotiation
// mechanism: a newer controller calls /status first and compares this
// against its own CONTROLLER_IPC_CONTRACT_VERSION before relying on any
// field a older server/worker pair might not have.

import { PROVISIONING_STATUS_SCHEMA, type ProvisioningStatus } from "@loombre/provisioning";
import { PROCESS_INFO_SCHEMA, type ProcessInfo } from "./process-info.js";

export interface VersionInfo {
  ipcContractVersion: number;
}

export const VERSION_INFO_SCHEMA = {
  type: "object",
  additionalProperties: false,
  required: ["ipcContractVersion"],
  properties: {
    ipcContractVersion: { type: "integer", minimum: 1 },
  },
} as const;

export interface IpcStatusResponse extends VersionInfo {
  server: ProcessInfo;
  worker: ProcessInfo;
  /** null while the server is not in a state that serves the web client. */
  webUrl: string | null;
  provisioning: ProvisioningStatus;
}

export const IPC_STATUS_RESPONSE_SCHEMA = {
  type: "object",
  additionalProperties: false,
  required: ["ipcContractVersion", "server", "worker", "webUrl", "provisioning"],
  properties: {
    ipcContractVersion: { type: "integer", minimum: 1 },
    server: PROCESS_INFO_SCHEMA,
    worker: PROCESS_INFO_SCHEMA,
    webUrl: { type: ["string", "null"] },
    provisioning: PROVISIONING_STATUS_SCHEMA,
  },
} as const;
